presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -447,6 +448,28 @@ private void processNewSources()
         }
     }
 
+    private void addBlockedOperator(List<Operator> blockedOperators, List<ListenableFuture<?>> blockedFutures, Operator operator, ListenableFuture<?> blockedFuture)
+    {
+        if (!blockedOperators.isEmpty() && getLast(blockedOperators) == operator) {
+            // We don't want to duplicate operators on the list. Just update its "entry" with new blocked future
+            blockedFutures.set(blockedFutures.size() - 1, blockedFuture);
+            return;
+        }
+
+        blockedOperators.add(operator);
+        blockedFutures.add(blockedFuture);
+    }
+
+    private static boolean allBlocked(List<ListenableFuture<?>> blockedFutures)
```

Comment:
I think we can live without this optimization. The code is long and complicated without it enough.

Reply:
Well, when there are a few operators in a driver, some time passes between when we first consulted the future and now, when we construct the `firstFinishedFuture`. `firstFinishedFuture` has visible overhead (it registers a listener in all futures), this is why i wanted to re-check the futures before calling `firstFinishedFuture`. Also, this is more in the spirit of the code before my change: we asked operators anew whether they are blocked.

Reply:
Ok. I am fine with that. Just do not see this necessary.